src/pydocstyle/checker.py

```diff
@@ -390,6 +411,174 @@ def check_starts_with_this(self, function, docstring):
             if first_word.lower() == 'this':
                 return violations.D404()
 
+    @staticmethod
+    def _get_leading_words(line):
+        """Return any leading set of words from `line`.
+
+        For example, if `line` is "  Hello world!!!", returns "Hello world".
+        """
+        result = re("[A-Za-z ]+").match(line.strip())
+        if result is not None:
+            return result.group()
+
+    @staticmethod
+    def _is_a_docstring_section(context):
+        """Check if the suspected line is really a section.
+
+        This is done by checking the following conditions:
+        * Does the current line has a suffix after the suspected section name?
+        * Is the previous line not empty?
+        * Does the previous line end with a punctuation mark?
+
+        If so, this is probably not a real section name. For example:
+            '''Title.
+
+            Some part of the docstring that specifies what the function
+            returns. <----- Not a real section name.
+            '''
+        """
+        section_name_suffix = context.line.lstrip(context.section_name).strip()
+
+        punctuation = [',', ';', '.', '-', '\\', '/', ']', '}', ')']
+        prev_line_ends_with_punctuation = \
+            any(context.previous_line.strip().endswith(x) for x in punctuation)
+
+        return not (section_name_suffix != '' and not
+                    prev_line_ends_with_punctuation and not
+                    context.previous_line.strip() == '')
```

Comment:
There are about 3 layers of negation here. I had to write it down to understand (I hope) what's happening. How about:

```
is_section = (not section_name_suffix or 
              prev_line_ends_with_punctuation or
              is_blank(context.previous_line)

return is_section
```

Also, the docstring should be clearer that *all* conditions should be satisfied for the line to *not* be considered a section name.